### PR TITLE
Fix `lr_finder` suggesting too high learning rates

### LINKED ISSUE #6471: Learning rate finder broke on master
## 🐛 Bug

Learning rate finder stopped working on master branch ([latest commit tested](https://github.com/PyTorchLightning/pytorch-lightning/commit/d1db604c61edb3142960346fbcbbfe74250baeb7)).

### To Reproduce

*Note: the BoringModel was not running so I used an other model from the docs*

See [this colab](https://colab.research.google.com/drive/1U2-S_5g1Crh-pj2JqFAPVO5nvb8xu5Dl?usp=sharing) to reproduce the bug.
It works on latest released version (1.2.3) but not when installing from latest master branch (it diverges).

### Expected behavior

Learning rate finder should converge

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Fixes #6471.

### Description of the changes
- [x] Stop training not only if `global_step+1` is equal to `max_steps` but also if it exceeds `max_steps`
- [x] Test lr finder ends at or before `num_training` steps
- [x] Test lr finder suggests between `max_lr` and `min_lr`

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/tuner/test_lr_finder.py [hunk 1]
```diff
@@ -295,3 +295,47 @@ def __init__(self, learning_rate=0.1, batch_size=2):
...
+            super().__init__()
+            self.save_hyperparameters()
+
+        def training_step_end(self, outputs):
+            assert self.global_step < num_training
+            return outputs
+
+    model = TestModel()
+    trainer = Trainer(default_root_dir=tmpdir)
+    num_training = 100
```
Comment (awaelchli) [inline tests/tuner/test_lr_finder.py:337]: can this number be reduced? we only need a few steps to know that it works right?

**Review by awaelchli — APPROVED** [review]

File: tests/tuner/test_lr_finder.py [hunk 2]
```diff
@@ -295,3 +295,47 @@ def __init__(self, learning_rate=0.1, batch_size=2):
...
+
+def test_lr_finder_ends_before_num_training(tmpdir):
+    """Tests learning rate finder ends before `num_training` steps."""
+    class TestModel(BoringModel):
+        def __init__(self, learning_rate=0.1):
+            super().__init__()
+            self.save_hyperparameters()
+
+        def training_step_end(self, outputs):
+            assert self.global_step < num_training
```
Comment (carmocca) [inline tests/tuner/test_lr_finder.py:332]: Unrelated to the changes in this PR, but a few questions about this

- Is the global_step increased during tuning?
- Is the optimizer state update during tuning?
- if yes, should it?

**Review by carmocca — APPROVED** [review]

File: tests/tuner/test_lr_finder.py (hunk 2 above)
Comment (akihironitta) [reply tests/tuner/test_lr_finder.py:332]: > Is the global_step increased during tuning?

Yes, but it gets reset after tuning [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/013756404be0b665f393955f731665e2c09f8a10/pytorch_lightning/tuner/lr_finder.py#L191).

> Is the optimizer state update during tuning?

Yes, It seems so when fitting a model with various learning rates [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/479122d17a05e075c6cb02a5f4af03888c36abd9/pytorch_lightning/tuner/lr_finder.py#L179), but this state also gets reset after tuning.

> if yes, should it?

As far as I understand [the original paper](https://arxiv.org/pdf/1506.01186.pdf), I think it should.

cc: @SkafteNicki

File: tests/tuner/test_lr_finder.py (hunk 1 above)
Comment (akihironitta) [reply tests/tuner/test_lr_finder.py]: Addressed.

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by s-rog — APPROVED** [review]

File: tests/tuner/test_lr_finder.py (hunk 2 above)
Comment (SkafteNicki) [reply tests/tuner/test_lr_finder.py:330]: agree with @akihironitta observations

**Review by SkafteNicki — APPROVED** [review]
